cpp/src/cache/image_cache_per_process.cpp

```diff
@@ -53,19 +55,38 @@ struct PerProcessImageCacheItem
     std::shared_ptr<ImageCacheValue> value;
 };
 
-PerProcessImageCacheValue::PerProcessImageCacheValue(void* data, uint64_t size, void* user_obj)
-    : ImageCacheValue(data, size, user_obj){};
+PerProcessImageCacheValue::PerProcessImageCacheValue(void* data,
+                                                     uint64_t size,
+                                                     void* user_obj,
+                                                     const cucim::io::DeviceType device_type)
+    : ImageCacheValue(data, size, user_obj, device_type){};
+
 PerProcessImageCacheValue::~PerProcessImageCacheValue()
 {
     if (data)
     {
-        cucim_free(data);
+        switch (device_type)
+        {
+        case io::DeviceType::kCPU:
+            cucim_free(data);
+            break;
+        case io::DeviceType::kCUDA: {
+            cudaError_t cuda_status;
+            CUDA_TRY(cudaFree(data));
+            break;
+        }
+        case io::DeviceType::kPinned:
+        case io::DeviceType::kCPUShared:
+        case io::DeviceType::kCUDAShared:
+            fmt::print(stderr, "Device type {} is not supported!\n", device_type);
+            break;
```

Comment:
Do we intend to cover these in the future? If so, should we raise an issue?

Note: This comes up again below.

Reply:
Created issue:
- https://github.com/rapidsai/cucim/issues/210